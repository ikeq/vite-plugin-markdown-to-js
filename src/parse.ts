import marked, { MarkedOptions } from 'marked';
import parseMatter from 'gray-matter';

type CodeFn = (scripts: CodeBlock[], html: CodeBlock[]) => string;

export interface CodeBlock<T = string | CodeFn> {
  type: string;
  skipCompile?: true;
  code: T;
  attrs?: string;
  transform?: (source: string) => string;
  [name: string]: any;
}

export interface LangTransform {
  lang: string;
  tokenize?: (tokens: marked.TokensList) => marked.TokensList;
  transform: (
    code: string,
    md: string,
    env: { matter: any; path: string }
  ) => CodeBlock | CodeBlock[];
}

export function parse(
  raw: string,
  transforms: LangTransform[],
  path: string,
  markedOptions?: MarkedOptions
) {
  const { data: matter, content } = parseMatter(raw);
  const tokens = marked.lexer(content);
  const html: CodeBlock[] = [];
  const style: CodeBlock[] = [];
  const script: CodeBlock[] = [];
  // function type is delayed
  const scriptFn: CodeBlock<CodeFn>[] = [];

  transforms?.forEach((tms) => {
    if (tms.tokenize) {
      tms.tokenize(tokens);
    }
  });

  tokens.forEach((token) => {
    if (token.type === 'space') return;
    if (token.type !== 'code' || !transforms) {
      html.push({ type: 'html', code: token.raw });
      return;
    }

    const transform = transforms.find((t) => t.lang === token.lang);

    if (!transform) {
      html.push({ type: 'html', code: token.raw });
      return;
    }

    const transformed = transform.transform(token.text, token.raw, { matter, path });

    transformed.flat(2).forEach((td: CodeBlock) => {
      if (td.type !== 'script') {
        (td.type === 'style' ? style : html).push(td);
        return;
      }
      (typeof td.code === 'function' ? scriptFn : script).push(td);
    });
  });

  if (markedOptions) {
    marked.setOptions(markedOptions);
  }

  const getAttrs = (...args: CodeBlock[][]) =>
    (args.flat(2).find((i) => i.attrs) || {}).attrs || '';

  return {
    script:
      `<script${getAttrs(script, scriptFn)}>\n` +
      [...script.map((i) => i.code), ...scriptFn.map((i) => i.code(script, html))].join('\n') +
      `\n</script>`,

    // comes last so that it can be modified by script
    html: html
      .map((i) => {
        if (!i.code) return '';
        if (i.skipCompile) return i.code;

        const ret = marked(i.code as string);
        return i.transform ? i.transform(ret) : ret;
      })
      .join(''),

    style: style.length
      ? `<style${getAttrs(style)}>\n${style.map((s) => s.code).join('\n')}\n</style>`
      : undefined,
    matter,
  };
}
