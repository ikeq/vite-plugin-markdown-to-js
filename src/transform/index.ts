export * from './imports';
export * from './vue';
