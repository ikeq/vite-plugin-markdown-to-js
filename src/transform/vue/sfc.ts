const tagOpeningRe = /\<((?:[a-z]+)(?:-[a-z]+)?)\s?/;

function parseTag(content: string, tag: string) {
  const matcher = {
    openingRe: new RegExp(`\\<${tag}([\\s\\S]*?)\\>`),
    closingRe: new RegExp(`\\<\\/${tag}\\>`)
  };
  let remaining = content;
  let pointer = -1;
  let indics = [];
  let attrs = '';

  while (remaining) {
    const opening = matcher.openingRe.exec(remaining);
    // none matching of open tag
    if (pointer < 0) {
      // malformed
      if (!opening) return null;
      // opening
      pointer += 1;
      indics[0] = opening.index + opening[0].length;
      attrs = opening[1];

      remaining = remaining.slice(opening.index + opening[0].length);
      continue;
    }

    // nesting
    if (opening && !matcher.closingRe.test(remaining.slice(0, opening.index))) {
      pointer += 1;

      remaining = remaining.slice(opening.index + opening[0].length);
      continue;
    } else {
      const closing = matcher.closingRe.exec(remaining);
      // malformed
      if (!closing) return null;

      pointer -= 1;
      remaining = remaining.slice(closing.index + closing[0].length);

      // closing
      if (pointer === -1) {
        indics[1] = content.length - remaining.length - closing[0].length;
        break;
      }
    }
  }

  return {
    code: content.slice(...indics),
    tag,
    attrs,
    remaining
  };
}

export function parseSFC<T>(raw: string): { code: string; tag: string & T; attrs: string }[] {
  const fn = (raw: string, ret: any = []) => {
    const group = tagOpeningRe.exec(raw);
    if (!group || !group[1]) return [];

    const block = parseTag(raw, group[1] as any);
    if (block) {
      ret.push({
        tag: block.tag,
        code: block.code,
        attrs: block.attrs
      });
      if (block.remaining) {
        fn(block.remaining, ret);
      }
    }

    return ret;
  }
  const blocks = fn(raw) as ReturnType<typeof parseSFC>;

  // merging
  const ret: Record<string, any> = {};

  blocks.forEach((i) => {
    if (!ret[i.tag]) {
      const attrs = i.attrs.trim();
      ret[i.tag] = {
        tag: i.tag,
        attrs: attrs ? ` ${attrs}` : '',
        code: ''
      };
    }
    ret[i.tag].code = ret[i.tag].code + '\n' + i.code.trim();
  });

  return Object.values(ret);
}
