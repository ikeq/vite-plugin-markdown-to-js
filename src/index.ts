import { MarkedOptions } from 'marked';
import { PluginOption } from 'vite';
import { LangTransform, parse } from './parse';

interface Options {
  transforms: LangTransform[];
  render?: (output: ReturnType<typeof parse>, env: { path: string }) => string;
  markedOptions?: MarkedOptions;
}

export * from './transform';

export { parse };

export default function vitePluginMarkdownToJs(options: Options): PluginOption {
  const transforms = options.transforms || [];
  const render = options.render
    || ((output) => `export default ${JSON.stringify(output.html + output.style + output.script)};`);

  function markdownToJs(raw: string, id: string) {
    return render(parse(raw, transforms, options.markedOptions), { path: id });
  }

  return {
    name: 'vite-plugin-markdown-to-js',
    enforce: 'pre',
    transform(raw, id) {
      if (id.endsWith('.md')) {
        try {
          return markdownToJs(raw, id);
        } catch (e) {
          this.error(e);
        }
      }
    },
    async handleHotUpdate(ctx) {
      if (ctx.file.endsWith('.md')) {
        const defaultRead = ctx.read;
        ctx.read = async function () {
          return markdownToJs(await defaultRead(), ctx.file);
        }
      }
    },
  }
};
