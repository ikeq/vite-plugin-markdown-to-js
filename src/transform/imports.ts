import camelCase from 'lodash/camelCase';
import { LangTransform } from '../parse';

const rSplit = /[\n\s,]/;
const rExt = /\.\w+$/;

/**
 * @example
 * ```md
 * ```imports
 * module1, module2, module3
 * module4
 * ```
 * ```
 */
export function transformImports({
  base,
  defaultPrefix,
}: { base?: string | ((path: string) => string); defaultPrefix?: string } = {}): LangTransform {
  const baseFn = (() => {
    if (typeof base === 'function') return base;
    if (typeof base === 'string') {
      if (base && !base.endsWith('/')) {
        base = `${base}/`;
      }
      return (path: string) => `${base || ''}${path}`;
    }
    return (path: string) => path;
  })();

  return {
    lang: 'imports',
    transform(src) {
      const imports = [...new Set(src.split(rSplit).filter((i) => i))];
      const importNames = imports.map((i) =>
        defaultPrefix ? camelCase(`${defaultPrefix} ${i}`) : i
      );

      return [
        {
          type: 'imports',
          imports: importNames,
          code: '',
        },
        {
          type: 'script',
          imports: importNames,
          code: imports
            .map((i, index) => {
              const path = baseFn(`${i}${rExt.test(i) ? '' : '.md'}`);
              return `import ${importNames[index]} from "${path}";`;
            })
            .join('\n'),
        },
      ];
    },
  };
}
