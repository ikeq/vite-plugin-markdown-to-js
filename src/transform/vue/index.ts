import kebabCase from 'lodash/kebabCase';
import { CodeBlock, LangTransform } from '../../parse';
import { parseSFC } from './sfc';


const defaultRenders: Record<'template' | 'script' | 'style', (code: string, meta: any, env: Parameters<LangTransform['transform']>[2]) => string> = {
  template: code => code.trim(),
  style: code => code.trim(),
  script: (code, { imports }) => {
    code = code.trim();
    if (!code) {
      return `export default { components: { ${imports || ''} } };`;
    }
    return code.replace('export default {', `export default {\n components: { ${imports || ''} },`);
  },
};

/**
 * @example
 * ```md
 * ```vue
 * <template>
 * </template>
 * <script>
 * export default {}
 * </script>
 * ```
 * ```
 */
export function transformVue(options: {
  importsAsComponents?: boolean | ((imports: string[]) => string);
  renders?: typeof defaultRenders
} = {}): LangTransform {
  return {
    lang: 'vue',
    tokenize(tokens) {
      if (!tokens.some((token) => token.type === 'code' && token.lang === 'vue')) {
        tokens.push({ type: 'code', lang: 'vue', text: '<script></script>', raw: '' });
      }

      return tokens;
    },
    transform(src, md, env) {
      const renders: typeof options.renders = {
        template: options.renders?.template || defaultRenders.template,
        style: options.renders?.style || defaultRenders.style,
        script: options.renders?.script || defaultRenders.script,
      };
      const blocks = parseSFC<'template' | 'script' | 'style'>(src);
      const ret: CodeBlock[] = [{
        type: 'html',
        code: md.replace('```vue', '```html'),
        transform: source => `<!-- langvue -->${source.replace(/^<pre>/, '<pre v-pre>')}<!-- endlangvue -->`,
      }];

      if (!blocks.length) return ret;

      const tags = blocks.reduce((map, i) => {
        return {
          ...map,
          [i.tag]: { ...i }
        };
      }, {} as Record<typeof blocks[0]['tag'], typeof blocks[0]>);

      if (!tags.script) {
        tags.script = {
          tag: 'script',
          code: '',
          attrs: '',
        };
      }
      if (!tags.template) {
        tags.template = {
          tag: 'template',
          code: '',
          attrs: '',
        };
      }
      if (!tags.style) {
        tags.style = {
          tag: 'style',
          code: '',
          attrs: '',
        };
      }

      Object.keys(tags).forEach((key) => {
        const tagName = key as keyof typeof tags;
        const block = tags[tagName];
        const renderer = renders[tagName];
        const code: CodeBlock['code'] = tagName === 'script' && options.importsAsComponents
          ? (scripts, html) => {
            const imports = scripts.map((i) => i.imports || []).flat(2);

            html.forEach((h) => {
              // consume imports block
              if (h.type === 'imports' && h.imports) {
                h.type = 'html';
                h.skipCompile = true;
                h.code =
                  typeof options.importsAsComponents === 'function'
                    ? options.importsAsComponents(h.imports)
                    : (h.imports.map((i: string) => `<${kebabCase(i)}/>`).join('') as string);
              }
            });

            return renderer(block.code, { imports }, env)
          }
          : renderer(block.code, {}, env);

        ret.unshift({ type: tagName, attrs: block.attrs, code, skipCompile: true });
      });

      return ret;
    },
  };
}
