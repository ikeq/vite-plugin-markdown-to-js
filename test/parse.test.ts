import { parse } from '../src/parse';
import { transformImports } from '../src/transform/imports';
import { transformVue } from '../src/transform/vue';

const md = `
# title

asdasdasd

asdasdasdasd
`;

const mdWithMatter = `---
foo: bar
---

# title

asdasdasd

asdasdasdasd
`;

const mdWithImports = `
# title

\`\`\`imports
a, b, c
\`\`\`

asdasdasdasd
`;

const mdWithVue = `
# title

\`\`\`vue
<template>
<div>
<template>{{ name }}</template>
</div>
</template>
<script>
export default {
  name: 'test',
  data() {
    return {
      name: 'foo'
    }
  }
}
</script>
<style>
.test { color: red; }
</style>
\`\`\`

asdasdasdasd
`;

const mdWithImportsVue = `
# title

\`\`\`imports
a, b, c
\`\`\`

\`\`\`vue
<template>
<div>
<template>asd</template>
</div>
</template>
<script lang="ts">
export default { name: 'test' }
</script>
<style lang="less" scoped>
.test { color: red; }
</style>
\`\`\`

asdasdasdasd
`;

describe('parse', () => {
  it('basic', () => {
    expect(parse(md)).toMatchSnapshot();
  });

  it('with matter', () => {
    expect(parse(mdWithMatter)).toMatchSnapshot();
  });

  it('transform imports', () => {
    expect(parse(mdWithImports, [
      transformImports()
    ])).toMatchSnapshot();
  });

  it('transform vue', () => {
    expect(parse(mdWithVue, [
      transformVue()
    ])).toMatchSnapshot();
  });

  it('transform imports & vue & importsAsComponents:true', () => {
    expect(parse(mdWithImportsVue, [
      transformImports(),
      transformVue({ importsAsComponents: true })
    ])).toMatchSnapshot();
  });
});
