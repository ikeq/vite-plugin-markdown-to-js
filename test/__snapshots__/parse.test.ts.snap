// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse basic 1`] = `
Object {
  "html": "<h1 id=\\"title\\">title</h1>
<p>asdasdasd</p>
<p>asdasdasdasd</p>
",
  "matter": Object {},
  "script": "<script>
export const matter = {};
</script>",
  "style": undefined,
}
`;

exports[`parse transform imports & vue & importsAsComponents:true 1`] = `
Object {
  "html": "<h1 id=\\"title\\">title</h1>
<a/><b/><c/>
<div>
<template>asd</template>
</div><pre><code class=\\"language-html\\">&lt;template&gt;
&lt;div&gt;
&lt;template&gt;asd&lt;/template&gt;
&lt;/div&gt;
&lt;/template&gt;
&lt;script lang=&quot;ts&quot;&gt;
export default { name: &#39;test&#39; }
&lt;/script&gt;
&lt;style lang=&quot;less&quot; scoped&gt;
.test { color: red; }
&lt;/style&gt;
</code></pre>
<p>asdasdasdasd</p>
",
  "matter": Object {},
  "script": "<script lang=\\"ts\\">
import a from \\"a.md\\";
import b from \\"b.md\\";
import c from \\"c.md\\";

export default {
 components: { a,b,c }, name: 'test' }
export const matter = {};
</script>",
  "style": "<style lang=\\"less\\" scoped>

.test { color: red; }
</style>",
}
`;

exports[`parse transform imports 1`] = `
Object {
  "html": "<h1 id=\\"title\\">title</h1>
<p>asdasdasdasd</p>
",
  "matter": Object {},
  "script": "<script>
import a from \\"a.md\\";
import b from \\"b.md\\";
import c from \\"c.md\\";
export const matter = {};
</script>",
  "style": undefined,
}
`;

exports[`parse transform vue 1`] = `
Object {
  "html": "<h1 id=\\"title\\">title</h1>

<div>
<template>{{ name }}</template>
</div><pre><code class=\\"language-html\\">&lt;template&gt;
&lt;div&gt;
&lt;template&gt;&#x7B;&#x7B; name &#x7D;&#x7D;&lt;/template&gt;
&lt;/div&gt;
&lt;/template&gt;
&lt;script&gt;
export default {
  name: &#39;test&#39;,
  data() {
    return {
      name: &#39;foo&#39;
    }
  }
}
&lt;/script&gt;
&lt;style&gt;
.test { color: red; }
&lt;/style&gt;
</code></pre>
<p>asdasdasdasd</p>
",
  "matter": Object {},
  "script": "<script>

export default {
  name: 'test',
  data() {
    return {
      name: 'foo'
    }
  }
}
export const matter = {};
</script>",
  "style": "<style>

.test { color: red; }
</style>",
}
`;

exports[`parse with matter 1`] = `
Object {
  "html": "<h1 id=\\"title\\">title</h1>
<p>asdasdasd</p>
<p>asdasdasdasd</p>
",
  "matter": Object {
    "foo": "bar",
  },
  "script": "<script>
export const matter = {\\"foo\\":\\"bar\\"};
</script>",
  "style": undefined,
}
`;
